### ISSUE TITLE
Failed To Write Player data

### ISSUE THREAD

**Issue by Cpl-Jordan** [CleverRaven/Cataclysm-DDA#48291]
# Describe the bug
The game is unable to save data.

# Steps To Reproduce
1.) Make a new world
2.) Start a new character
3.) Save and quit
4.) Fail message appears 

# Expected behavior
Being able to save my progress

# Versions and configuration
- OS: Windows
    - OS Version: 10.0 2004
- Game Version: 0.E-10255-g6b8dd00 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Bionic Professions [package_bionic_professions]
]

**Comment by ZhilkinSerg** [comment]
Install your game to a folder you have write access to.

**Comment by RDru** [comment]
Updated to 0.E-10255-g6b8dd00 (tiles)  / 11557 and also got this bug
![1](https://user-images.githubusercontent.com/26722636/113182336-12c87500-9253-11eb-822c-e851a6f6f3d8.png)

**Comment by Cpl-Jordan** [comment]
> Install your game to a folder you have write access to.

I don't think that is the problem. I have the game installed in the same folder as always have. I use CDDA Game Launcher to update and keep me up to date. I haven't had this problem until Build 11557.

**Comment by Lamandus** [comment]
Can confirm! 11556 is saving normally, while 11557 is erroring out on me, too.
The only update in 11557 is #48249

**Comment by Lamandus** [comment]
nvm also 11556 is throwing that error.
![grafik](https://user-images.githubusercontent.com/33199510/113189399-75257380-925b-11eb-896a-812d67580ca9.png)
- OS: Windows
    - OS Version: 10.0 2004
- Game Version: 0.E-10253-g5715a33 [32-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    meins [mine],
    Nonperishable Overhaul [Nonperishable_Overhaul],
    Fewer Farms [Fewer_Farms],
    SpeedyDex [speedydex],
    Stats Through Skills [StatsThroughSkills]
]

**Comment by BrettDong** [comment]
Fails here:
<img width="1552" src="https://user-images.githubusercontent.com/21075502/113195879-d08e3c00-9230-11eb-9f24-a94458ea7c28.png">

**Comment by BrettDong** [comment]
Regression from #45316

**Comment by BrettDong** [comment]
If I understand the code correctly, the problem here is that the new `activity_history.average_activity()` returns a continuous value, but `log_activity_level` expects a discrete value from `{1, 2, 4, 6, 8, 10}` defined in `game_constants.h` line 165:
```c++
constexpr float NO_EXERCISE = 1.0f;
constexpr float LIGHT_EXERCISE = 2.0f;
constexpr float MODERATE_EXERCISE = 4.0f;
constexpr float BRISK_EXERCISE = 6.0f;
constexpr float ACTIVE_EXERCISE = 8.0f;
constexpr float EXTRA_EXERCISE = 10.0f;
```
So `save_activity` encounters some value it doesn't recognise (`0.0f` in the screenshot above) and doesn't know how to serialise.

Ping @kevingranade

**Comment by wapcaplet** [comment]
Confirmed here as well in 0.E-10255-g6b8dd0058f, simply creating a new world and saving, I get the "Failed to write player data" error.

**Comment by shmup** [comment]
this bug surfaces on linux too

**Comment by actual-nh** [comment]
I think I need to hold back on revisions to #47653 (due to #45316)... 0.0f? Lower than NO_ACTIVITY? Or is this also rescaled or something?

**Comment by zebbs** [comment]
This bug affects Linux, too, with the exact same error message. Sadly, it renders the game 100% unplayable. Good luck, devs.

**Comment by zebbs** [comment]
> > Sadly, it renders the game 100% unplayable. Good luck, devs.
> 
> There are thousands of versions that are not affected by this bug and playable, you know.

You can't even save and quit.  You have to switch out of the game and manually `kill` it.  Thanks for your brilliant insights, though.  CDDA is lucky to have someone with your skills on their team.

**Comment by ZhilkinSerg** [comment]
> > > Sadly, it renders the game 100% unplayable. Good luck, devs.
> > 
> > 
> > There are thousands of versions that are not affected by this bug and playable, you know.
> 
> You can't even save and quit. You have to switch out of the game and manually `kill` it. Thanks for your brilliant insights, though. CDDA is lucky to have someone with your skills on their team.

Have you started doing backups already?

**Comment by zebbs** [comment]
> > > > Sadly, it renders the game 100% unplayable. Good luck, devs.
> > > 
> > > 
> > > There are thousands of versions that are not affected by this bug and playable, you know.
> > 
> > 
> > You can't even save and quit. You have to switch out of the game and manually `kill` it. Thanks for your brilliant insights, though. CDDA is lucky to have someone with your skills on their team.
> 
> Have you started doing backups already?

Have you started learning social skills?

**Comment by Mom-Bun** [comment]
> > > Sadly, it renders the game 100% unplayable. Good luck, devs.
> > 
> > 
> > There are thousands of versions that are not affected by this bug and playable, you know.
> 
> You can't even save and quit. You have to switch out of the game and manually `kill` it. Thanks for your brilliant insights, though. CDDA is lucky to have someone with your skills on their team.

It is called "experimental" for a reason, if you want a guarantee that you'll have a playable game, simply check out stable, that is guaranteed to work and be stable!

However, since this is experimental, and old builds still exist right now, and the current one is broken, simply download an old build, and practice safety in either backing up, or not updating to a new experimental every day.

**Comment by Kragh50** [comment]
> > > Sadly, it renders the game 100% unplayable. Good luck, devs.
> > 
> > 
> > There are thousands of versions that are not affected by this bug and playable, you know.
> 
> You can't even save and quit. You have to switch out of the game and manually `kill` it. Thanks for your brilliant insights, though. CDDA is lucky to have someone with your skills on their team.

I mean, he could have been nicer about his answer, but he's not wrong. If you have automatic backup when updating (which you should do when playing experimental), you can just load the backup and update later when it's fixed.

**Comment by zebbs** [comment]
> > > There are thousands of versions that are not affected by this bug and playable, you know.
> > 
> > 
> > You can't even save and quit. You have to switch out of the game and manually `kill` it. Thanks for your brilliant insights, though. CDDA is lucky to have someone with your skills on their team.
> 
> It is called "experimental" for a reason, if you want a guarantee that you'll have a playable game, simply check out stable, that is guaranteed to work and be stable!
> 
> However, since this is experimental, and old builds still exist right now, and the current one is broken, simply download an old build, and practice safety in either backing up, or not updating to a new experimental every day.

I _have_ playing "other versions" for the last 8 months: Bright Nights.  I seem to have picked a bad time to check out Kevin's CDDA fork again.

As for backups, I'd love to be able to save a CDDA game at all.  Good job, assuming I've been riding your train wreck this whole time.

**Comment by actual-nh** [comment]
> I _have_ playing "other versions" for the last 8 months: Bright Nights. I seem to have picked a bad time to check out Kevin's CDDA fork again.

Well, yes, you did, I agree. I suggest if you're going to try a game out very occasionally, you might want to do so when it has a stable release.

**Comment by BrettDong** [comment]
Build 11552 is a relatively safe version to play at the moment.

**Comment by ZhilkinSerg** [comment]
> 
> 
> > > > There are thousands of versions that are not affected by this bug and playable, you know.
> > > 
> > > 
> > > You can't even save and quit. You have to switch out of the game and manually `kill` it. Thanks for your brilliant insights, though. CDDA is lucky to have someone with your skills on their team.
> > 
> > 
> > It is called "experimental" for a reason, if you want a guarantee that you'll have a playable game, simply check out stable, that is guaranteed to work and be stable!
> > However, since this is experimental, and old builds still exist right now, and the current one is broken, simply download an old build, and practice safety in either backing up, or not updating to a new experimental every day.
> 
> I _have_ playing "other versions" for the last 8 months: Bright Nights. I seem to have picked a bad time to check out Kevin's CDDA fork again.
> 
> As for backups, I'd love to be able to save a CDDA game at all. Good job, assuming I've been riding your train wreck this whole time.

So, you are still not doing any backups. If that did not taught you anything, go play more BN probably.

**Comment by BrettDong** [comment]
> accusing people of having write permissions set incorrectly

This is a natural intuition when seeing this bug report at the first glance and I don't see it inappropriate in any aspect.

**Comment by actual-nh** [comment]
> Are you people developers, or trolls?

Umm.... are you a troll? Given that you're laughing at your own posts (see above)...

**Comment by actual-nh** [comment]
> > accusing people of having write permissions set incorrectly
> 
> This is a natural intuition when seeing this bug report at the first glance and I don't see it inappropriate in any aspect.

True, although one would hope that write permissions being set incorrectly would result in an error message that's more specific. Admittedly, this can be due to OS bugs.

**Comment by actual-nh** [comment]
> @actual-nh He has been banned from the official discord server, it is safe to say that he is indeed a troll.

@meelock: I've had a problem myself with getting registered with the discord server - the account registration protection mechanism doesn't like my email address, and assumes that I've got a Facebook/whatever account if I'm legit, which isn't the case. Any suggestions?

**Comment by meelock** [comment]
@actual-nh That's a bit strange indeed, I have encountered a similar issue with registering for a image hosting website... I think you might need to use a gmail account for discord, I don't remember it needing a facebook account though.  I think the discord server is worth joining, but there is no particular advantage of joining.

**Comment by esotericist** [comment]
> > @actual-nh He has been banned from the official discord server, it is safe to say that he is indeed a troll.
> 
> @meelock: I've had a problem myself with getting registered with the discord server - the account registration protection mechanism doesn't like my email address, and assumes that I've got a Facebook/whatever account if I'm legit, which isn't the case. Any suggestions?

given your established presence within the project, i think we can work something out if you contact a member of the moderation team while in the pre-authorization lobby.

**Comment by ZhilkinSerg** [comment]
> 
> 
> > So, you are still not doing any backups. If that did not taught you anything, go play more BN probably.
> 
> I can't do any backups _in CDDA_ because the game is no longer capable of saving at all. BrettDong has got this bug almost completely figured out, meanwhile you MomBun are acting like cultists, accusing people of having write permissions set incorrectly, or failing to have backups. Are you people developers, or trolls?

There are thousands of CDDA versions capable of saving everything correctly - use one of them.

**Comment by ZhilkinSerg** [comment]
> 
> 
> > > accusing people of having write permissions set incorrectly
> > 
> > 
> > This is a natural intuition when seeing this bug report at the first glance and I don't see it inappropriate in any aspect.
> 
> True, although one would hope that write permissions being set incorrectly would result in an error message that's more specific. Admittedly, this can be due to OS bugs.

There are just a few custom error messages now (writing/opening/moving file failed). Exceptions сould be more meaningful after we switch to `std:filesystem`.

**Comment by actual-nh** [comment]
> > I've had a problem myself with getting registered with the discord server - the account registration protection mechanism doesn't like my email address, and assumes that I've got a Facebook/whatever account if I'm legit, which isn't the case. Any suggestions?
> 
> given your established presence within the project, i think we can work something out if you contact a member of the moderation team while in the pre-authorization lobby.

Thanks! I will see about it sometime soon if possible; it's coming up on the end of the semester, and I admit to having procrastinated on some grading... sigh. I may need to hold off until after the semester is over.

> edit: also, it _should_ let you associate your github account with your discord account; the bot is supposed to be looking for associations between the discord account and other long-term internet presences. i would expect github to qualify.

I think it may have been because this was a few months back - KorGgenT was wondering if we could conference on the discord server about a PR I was helping someone with.

**Comment by Mom-Bun** [comment]
> > So, you are still not doing any backups. If that did not taught you anything, go play more BN probably.
> 
> I can't do any backups _in CDDA_ because the game is no longer capable of saving at all. BrettDong has got this bug almost completely figured out, meanwhile you MomBun are acting like cultists, accusing people of having write permissions set incorrectly, or failing to have backups. Are you people developers, or trolls?

Here, I'll spell it out for you:

Back up: "make a spare copy of data or a disk."

Now, since you weren't playing CDDA as you admitted, there is nothing to back up.

So the solution, (now I'm being nice here, and presuming you actually want a fix and just being incredibly obtuse here) is to _download an earlier build that allows saving and play on that one_

Or the other solution, if you actually aren't looking to fix this and just cause problems: stop purposefully trying be a troll and be a problem on the github here just like how you decided to get yourself banned from the development discord just to be a problem, and stop obsessing over DDA and basing your entire personality around hating it and the developers, and just play BNs and have fun.

**Comment by kevingranade** [comment]
This issue has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/cant-download-the-game/26177/4

**Comment by Moltenhead** [comment]
The issue can get temporarily fixed reverting 8bc8877. At least it worked on my part.

**Comment by actual-nh** [comment]
> The issue can get temporarily fixed reverting [8bc8877](https://github.com/CleverRaven/Cataclysm-DDA/commit/8bc8877fc48b621955840958a540c711f331a30a). At least it worked on my part.

That is something people can do until #48308 is fully functional (currently has a bug w/regard to multiple saves) and in master, yes.

**Comment by assumecroons** [comment]
The behavior happened for me on `0.E-10268-g7c4d41985d` with curses under Debian 10.9 (stable).

The temporary fix (`git revert -n 8bc8877` and then re-build) worked for me.
